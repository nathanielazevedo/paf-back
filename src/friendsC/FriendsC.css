/** @format */

.friends-list {
  width: 95%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  min-height: 95vh;
  margin-top: 55px;
  
}

.container-button {
  flex-grow: 1;
}

.add-icon {
  background-color: rgb(128, 125, 125);
  font-size: 20px;
  color: white;
  margin-right: 15px;
  transition: 1s;
  margin-top: 10px;
  border: none;
  padding: 5px;
  border-radius: 5px;
  padding-left: 10px;
  padding-right: 10px;
}

.add-icon:hover {
  font-size: 48px;
}
