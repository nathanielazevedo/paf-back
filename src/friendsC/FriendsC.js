/** @format */

import React, { useState, useEffect } from "react";
import "./FriendsC.css";
import Friend from "../friend/Friend";
import Form from "../form/Form";
import Paf from "../api.js";
import { Button } from "reactstrap";

//All API calls for friend info lives here

function FriendsC({ username }) {
  const [form, setForm] = useState(false);
  const [friends, setFriends] = useState([]);

  //TOGGLE ADD FRIEND FORM
  const openForm = () => {
    setForm(true);
  };
  const closeForm = () => {
    setForm(false);
  };

  //DELETE FRIEND
  async function deleteFriend(id) {
    await Paf.deleteFriend(id);
    let pos = friends
      .map((f) => {
        return f.id;
      })
      .indexOf(id);

    setFriends((data) => {
      data.splice(pos, 1);

      return [...data];
    });
  }

  // EDIT FRIEND
  async function editFriend(id, data) {
    let newFriend = await Paf.editFriend(id, data);
    let pos = friends
      .map((f) => {
        return f.id;
      })
      .indexOf(id);

    setFriends((datas) => {
      datas[pos] = newFriend;

      return [...datas];
    }); 
    return { success: true };

  }

  // ADD FRIEND
  const addFriend = async (formData) => {
    let res = await Paf.addFriend(formData);
    setFriends((data) => {
      if (data) {
        return [...data, res];
      } else {
        return [res];
      }
    });
    setForm(false);
    return { success: true };
  };

  // GET ALL FRIENDS
  useEffect(
    function () {
      async function getFriends() {
        let friends = await Paf.getAllFriends(username);
        setFriends(friends);
      }
      getFriends();
    },
    [username]
  );

  //GENERATE FRIENDS
  let allFriends = friends
    ? friends.map((f, i) => {
        return (
          <Friend
            name={f.name}
            description={f.description}
            id={f.id}
            editFunc={editFriend}
            deleteFunc={() => deleteFriend(f.id)}
            key={i}
          />
        );
      })
    : null;

  return (
    <>
      <ul className="friends-list">
        {allFriends}
        <div className="container-button">
          <Button onClick={openForm}>ADD FRIEND</Button>
        </div>
      </ul>
      {form ? (
        <Form
          title="New Friend"
          inputs={["name", "description"]}
          func={addFriend}
          after={`/friends`}
          close={closeForm}
        />
      ) : null}
    </>
  );
}

export default FriendsC;
