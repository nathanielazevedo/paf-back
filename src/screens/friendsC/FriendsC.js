/** @format */

import React, { useState, useEffect } from "react";
import Friend from "../../components/friend/Friend";
import Form from "../../components/form/Form";
import Paf from "../../api.js";
import { Button } from "reactstrap";
import "./FriendsC.css";

//All API calls for FRIENDS are initiated here. (CRUD)

function FriendsC({ username }) {
  const [form, setForm] = useState(false);
  const [friends, setFriends] = useState([]);

  //TOGGLE ADD FRIEND FORM
  const openForm = () => {
    setForm(true);
  };
  const closeForm = () => {
    setForm(false);
  };

  //GET ALL. 
  useEffect(
    function () {
      async function getFriends() {
        let friends = await Paf.getAllFriends(username);
        setFriends(friends);
      }
      getFriends();
    },
    [username]
  );

  //CREATE
  const addFriend = async (formData) => {
    let res = await Paf.addFriend(formData);
    setFriends((data) => {
      if (data) {
        return [...data, res];
      } else {
        return [res];
      }
    });
    setForm(false);
    return { success: true };
  };

  //UPDATE
  async function editFriend(id, data) {
    let newFriend = await Paf.editFriend(id, data);
    let pos = friends
      .map((f) => {
        return f.id;
      })
      .indexOf(id);

    setFriends((datas) => {
      datas[pos] = newFriend;

      return [...datas];
    });
    return { success: true };
  };

  //DELETE
  async function deleteFriend(id) {
    await Paf.deleteFriend(id);
    let pos = friends
      .map((f) => {
        return f.id;
      })
      .indexOf(id);

    setFriends((data) => {
      data.splice(pos, 1);

      return [...data];
    });
  }

  //GENERATE FRIENDS
  let allFriends = friends
    ? friends.map((f, i) => {
        return (
          <Friend
            name={f.name}
            description={f.description}
            id={f.id}
            editFunc={editFriend}
            deleteFunc={() => deleteFriend(f.id)}
            key={i}
          />
        );
      })
    : null;

  //Returns list of all friends a user has created.
  return (
    <>
      <ul className="friends-list">
        {allFriends}
        <div className="container-button">
          <Button onClick={openForm}>ADD FRIEND</Button>
        </div>
      </ul>
      {form ? (
        <Form
          title="New Friend"
          inputs={["name", "description"]}
          func={addFriend}
          after={`/paf-front-end/friends`}
          close={closeForm}
        />
      ) : null}
    </>
  );
}

export default FriendsC;
