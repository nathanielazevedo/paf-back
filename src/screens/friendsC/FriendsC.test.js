/** @format */

import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react";
import { MemoryRouter } from "react-router";
import FriendsC from "./FriendsC";
import { UserProvider } from "../../testUtils";
import { rest } from "msw";
import { setupServer } from "msw/node";

//Overides API requests and returns these values for testing
const server = setupServer(
  rest.get("http://localhost:3001/friends/Nate", (req, res, ctx) => {
    return res(
      ctx.status(200),
      ctx.json([{ name: "james", description: "test", id: 1 }])
    );
  }),
  rest.post("http://localhost:3001/friends/undefined", (req, res, ctx) => {
    return res(
      ctx.status(200),
      ctx.json({ name: "newjames", description: "test", id: 1 })
    );
  }),
  rest.delete("http://localhost:3001/friends/undefined/1", (req, res, ctx) => {
    return res(ctx.status(200), ctx.json({ message: "success" }));
  }),
  rest.patch("http://localhost:3001/friends/undefined/1", (req, res, ctx) => {
    return res(
      ctx.status(200),
      ctx.json({ name: "edited", description: "test", id: 1 })
    );
  })
);

beforeAll(() => server.listen());
afterAll(() => server.close());
afterEach(() => server.resetHandlers());

it("renders without crashing", function () {
  let username = "Nate";
  render(
    <MemoryRouter>
      <UserProvider>
        <FriendsC username={username} />
      </UserProvider>
    </MemoryRouter>
  );
});

it("matches snapshot", function () {
  let username = "Nate";
  const { asFragment } = render(
    <MemoryRouter>
      <UserProvider>
        <FriendsC username={username} />
      </UserProvider>
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});

it("displays expected text", function () {
  let username = "Nate";
  const { getAllByText } = render(
    <MemoryRouter>
      <UserProvider>
        <FriendsC username={username} />
      </UserProvider>
    </MemoryRouter>
  );

  expect(getAllByText("ADD FRIEND")[0]).toBeInTheDocument();
});

it("can open the form", function () {
  let username = "Nate";
  const { getAllByText } = render(
    <MemoryRouter>
      <UserProvider>
        <FriendsC username={username} />
      </UserProvider>
    </MemoryRouter>
  );

  const add = getAllByText("ADD FRIEND")[0];
  fireEvent.click(add);

  expect(getAllByText("New Friend")[0]).toBeInTheDocument();
});

it("can close the form", async function () {
  let username = "Nate";

  const {
    getAllByText,
    queryByText,
    getByPlaceholderText,
    getByTestId,
    getByText,
  } = render(
    <MemoryRouter>
      <UserProvider>
        <FriendsC username={username} />
      </UserProvider>
    </MemoryRouter>
  );
  await waitFor(() => expect(queryByText("james")).toBeInTheDocument());

  const add = getAllByText("ADD FRIEND")[0];
  fireEvent.click(add);

  const closeBtn = getByTestId("close");
  fireEvent.click(closeBtn);

  expect(queryByText("New Friend")).not.toBeInTheDocument();
});

it("can fill form and submit", async function () {
  let username = "Nate";

  const {
    getAllByText,
    queryByText,
    getByPlaceholderText,
    getByTestId,
    getByText,
  } = render(
    <MemoryRouter>
      <UserProvider>
        <FriendsC username={username} />
      </UserProvider>
    </MemoryRouter>
  );
  await waitFor(() => expect(queryByText("james")).toBeInTheDocument());

  const add = getAllByText("ADD FRIEND")[0];
  fireEvent.click(add);

  const nameInput = getByPlaceholderText("name");
  const descriptionInput = getByPlaceholderText("description");
  const submitBtn = getByTestId("button");

  // fill out the form
  fireEvent.change(nameInput, { target: { value: "james" } });
  fireEvent.change(descriptionInput, { target: { value: "friend" } });
  fireEvent.click(submitBtn);
});

it("It loads from MOCK API", async function () {
  let username = "Nate";

  const {
    getAllByText,
    queryByText,
    getByPlaceholderText,
    getByTestId,
    getByText,
  } = render(
    <MemoryRouter>
      <UserProvider>
        <FriendsC username={username} />
      </UserProvider>
    </MemoryRouter>
  );

  await waitFor(() => expect(queryByText("james")).toBeInTheDocument());
});

it("clicking add friend displays form then submit shows new friend", async function () {
  let username = "Nate";

  const {
    getAllByText,
    queryByText,
    getByPlaceholderText,
    getByTestId,
    getByText,
  } = render(
    <MemoryRouter>
      <UserProvider>
        <FriendsC username={username} />
      </UserProvider>
    </MemoryRouter>
  );

  await waitFor(() => expect(queryByText("james")).toBeInTheDocument());

  const add = getAllByText("ADD FRIEND")[0];
  fireEvent.click(add);

  const nameInput = getByPlaceholderText("name");
  const descriptionInput = getByPlaceholderText("description");
  const submitBtn = getByTestId("button");

  // fill out the form
  fireEvent.change(nameInput, { target: { value: "newjames" } });
  fireEvent.change(descriptionInput, { target: { value: "friend" } });
  fireEvent.click(submitBtn);

  await waitFor(() => expect(getByText("newjames")).toBeInTheDocument());
});

it("clicking delete removes the friend", async function () {
  let username = "Nate";

  const {
    getAllByText,
    queryByText,
    getByPlaceholderText,
    getByTestId,
    getByText,
  } = render(
    <MemoryRouter>
      <UserProvider>
        <FriendsC username={username} />
      </UserProvider>
    </MemoryRouter>
  );
  let name;
  await waitFor(() => {
    expect(queryByText("james")).toBeInTheDocument();
    name = queryByText("james");
  });

  const deleteBtn = getByText("Delete Friend");

  fireEvent.click(deleteBtn);

  await waitFor(() => expect(name).not.toBeInTheDocument());
});

it("can edit a friend", async function () {
  let username = "Nate";

  const {
    getAllByText,
    queryByText,
    getByPlaceholderText,
    getByTestId,
    getByText,
  } = render(
    <MemoryRouter>
      <UserProvider>
        <FriendsC username={username} />
      </UserProvider>
    </MemoryRouter>
  );

  await waitFor(() => {
    expect(queryByText("james")).toBeInTheDocument();
  });

  const editBtn = getByText("Edit Friend");

  fireEvent.click(editBtn);

  const nameInput = getByPlaceholderText("name");
  const descriptionInput = getByPlaceholderText("description");
  const submitBtn = getByTestId("button");

  // fill out the form
  fireEvent.change(nameInput, { target: { value: "newjames" } });
  fireEvent.change(descriptionInput, { target: { value: "friend" } });
  fireEvent.click(submitBtn);

  await waitFor(() => expect(getByText("edited")).toBeInTheDocument());
});
