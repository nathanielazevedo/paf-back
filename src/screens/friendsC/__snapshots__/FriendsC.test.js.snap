// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <ul
    class="friends-list"
  >
    <div
      class="container-button"
    >
      <button
        class="btn btn-secondary"
        type="button"
      >
        ADD FRIEND
      </button>
    </div>
  </ul>
</DocumentFragment>
`;
